fdbclient/ManagementAPI.actor.cpp

```diff
@@ -465,6 +467,160 @@ bool isCompleteConfiguration(std::map<std::string, std::string> const& options)
+                                        std::map<std::string, std::string> newConfig,
+                                        bool creating) {
+	if (creating) {
+		if (newConfig.count(encryptionAtRestModeConfKey.toString()) != 0) {
+			EncryptionAtRestMode encryptMode = EncryptionAtRestMode::fromValueRef(
+			    ValueRef(newConfig.find(encryptionAtRestModeConfKey.toString())->second));
+			if (encryptMode.mode == EncryptionAtRestMode::DISABLED ||
+			    encryptMode.mode == EncryptionAtRestMode::CLUSTER_AWARE) {
+				// In these cases the tenantMode does not matter
+				return true;
+			}
+
+			// check if the tenant mode is being set during configure new (otherwise assume tenants are disabled)
+			TenantMode tenantMode = TenantMode::DISABLED;
+			if (newConfig.count(tenantModeConfKey.toString()) != 0) {
+				tenantMode = TenantMode::fromValue(ValueRef(newConfig.find(tenantModeConfKey.toString())->second));
+			}
+			TraceEvent(SevDebug, "EncryptAndTenantModes")
+			    .detail("EncryptMode", encryptMode.toString())
+			    .detail("TenantMode", tenantMode.toString());
+			if (tenantMode != TenantMode::REQUIRED) {
+				// For domain aware encryption only the required tenant mode is currently supported
+				TraceEvent(SevWarnAlways, "InvalidEncryptAndTenantConfiguration")
+				    .detail("EncryptMode", encryptMode.toString())
+				    .detail("TenantMode", tenantMode.toString());
+				return false;
+			}
+			return true;
+		}
+	} else {
+		if (newConfig.count(tenantModeConfKey.toString()) != 0) {
+			ASSERT(oldConfiguration.present());
+			// If the tenant mode is being changed then make sure it obeys the current cluster encryption state
+			EncryptionAtRestMode encryptMode = oldConfiguration.get().encryptionAtRestMode;
+
+			if (encryptMode.mode == EncryptionAtRestMode::DISABLED) {
+				// In this case the tenantMode does not matter
+				return true;
+			}
+			TenantMode oldTenantMode = oldConfiguration.get().tenantMode;
+			TenantMode newTenantMode = TenantMode::fromValue(newConfig.find(tenantModeConfKey.toString())->second);
+			TraceEvent(SevDebug, "EncryptAndTenantModes")
+			    .detail("EncryptMode", encryptMode.toString())
+			    .detail("OldTenantMode", oldTenantMode.toString())
+			    .detail("NewTenantMode", newTenantMode.toString());
+			if (encryptMode.mode == EncryptionAtRestMode::DOMAIN_AWARE && newTenantMode != TenantMode::REQUIRED) {
+				// For domain aware encryption only the required tenant mode is currently supported (changing to any
+				// other mode is disallowed)
+				TraceEvent(SevWarnAlways, "InvalidEncryptAndTenantConfiguration")
+				    .detail("EncryptMode", encryptMode.toString())
+				    .detail("OldTenantMode", oldTenantMode.toString())
+				    .detail("NewTenantMode", newTenantMode.toString());
+				return false;
+			}
+			if ((oldTenantMode == TenantMode::OPTIONAL_TENANT || oldTenantMode == TenantMode::DISABLED) &&
+			    newTenantMode == TenantMode::REQUIRED) {
```

Comment:
Once we follow comment at line 500 above, this condition is already covered. 

In summary, I think line 510 - 536 has a lot of potential for simplification (and code removal).

Reply:
I added the comment but i still don't think this condition is covered. In this case we are trying to switch from optional/disabled -> required when the encrypt mode is cluster_aware. Right now this should fail until we implement logic to see if the cluster contains any non tenant data (in which case we all this change to succeed)

Reply:
If we allow this to change then we may run into errors if there exists any non tenant data in the cluster

Reply:
For instance: below is an attempt to rewrite block to cover all possible cases (hopefully)

// Encryption mode && tenant_mode can't change after creation 
`if (encryptionMode == domain_aware ) {
  assert(oldTenantMode == required)
  return newTenantMode == oldTenantMode;
} else if (newTenantMode != oldTenantMode) {
   if (oldTenantMode != required) {
     assert(newTenantMode == required);
     return false;
   }
   return true;
}`

Reply:
Discussed offline